### ISSUE TITLE
must_change ignores from:, to: under Ruby 3

### ISSUE THREAD

**Issue by bobmazanec** [minitest/minitest-rails#246]
At c870f2e
Ruby 3.1.2
Rails 7.0.4
Minitest 5.16.3

`{from: ..., to: ...}` becomes the `message` passed to `assert_changes`, leaving its `from` & `to` defaulted to [`UNTRACKED`](https://github.com/rails/rails/blob/5bebef36ade9103b1e4ebcfdafcf0c7361f75c45/activesupport/lib/active_support/testing/assertions.rb#L8) (`Object.new`)

For example, changing `test/rails/active_support/test_expectations.rb` to include
```
[...]
class TestActiveSupportExpectations < ActiveSupport::TestCase
  def test_must_change
    counter = 99 # CHANGED THIS TO FORCE FAILURE

    value { counter += 1 }.must_change "counter", from: 0, to: 1
[...]
```

does not fail

Digging into the backtrace, I find

```
[...]
  #3    ActiveSupport::Testing::Assertions#assert_changes(expression="counter", message={:from=>2, :to=>0}, from=#<Object:0x...>, to=#<Object:0x...>, block=#<Proc:0x... ...) at [...]gems/activesupport-7.0.4/lib/active_support/testing/assertions.rb:179
  #4    Minitest::Expectation#must_change(args=["counter", {:from=>0, :to=>1}]) at [...]gems/minitest-5.16.3/lib/minitest/spec.rb:26
[...]
```

~I believe it's due to minitest's [`Module::infect_an_assertion`](https://github.com/minitest/minitest/blob/b5565c0c7ab2ce5ecf306487cab0f2abe5377d55/lib/minitest/spec.rb#L4) not knowing/accounting-for rails's [`ActiveSupport::Testing::Assertions.assert_changes`](https://github.com/rails/rails/blob/5bebef36ade9103b1e4ebcfdafcf0c7361f75c45/activesupport/lib/active_support/testing/assertions.rb#L175) default args~
```
def assert_changes(expression, message = nil, from: UNTRACKED, to: UNTRACKED, &block)
```
but wanted to start the discussion here in case others found it difficult to create a failing test with this (& possibly other "infected assertions"?)

**Comment by bobmazanec** [comment]
Apologies if this is off-base and/or just not helpful, but as I bump around the code and am learning how this all works, it kind of reminds me of the `args`/`kwargs` cases/handling in `Minitest::Mock::`{[`expect`](https://github.com/minitest/minitest/blob/b5565c0c7ab2ce5ecf306487cab0f2abe5377d55/lib/minitest/mock.rb#L91),[`method_missing`](https://github.com/minitest/minitest/blob/b5565c0c7ab2ce5ecf306487cab0f2abe5377d55/lib/minitest/mock.rb#L150)}

Because of `infect_an_assertion`'s optional `dont_flip` parameter, maybe a new method that takes `dont_flip:` and another (optional)  named argument so the caller can pass some info/hints about `meth`'s positional and/or named arguments (maybe just the # of positional ones?)?
e.g.,
`infect_an_assertion2 :assert_changes, :must_change, dont_flip: :block, positional_args: 1`

(Even if the direction has merit, can probably do better than `...assertion2` 🤷)

**Comment by bobmazanec** [comment]
In case it helps, I tried it with other version combinations

minitest-rails c870f2e, Rails 7.0.4
| Ruby | Original test<br>(`counter = 0`) | Modified test<br>(`counter = 99`)
| - | - | -
| 3.1.2 | Passes with no warnings | Passes with no warnings
| 3.0.4 | Passes with no warnings | Passes with no warnings
| 2.7.6 | Passes with<br>`[...]/minitest-5.16.3/lib/minitest/spec.rb:26: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call`<br>`[...]/gems/activesupport-7.0.4/lib/active_support/testing/assertions.rb:175: warning: The called method 'assert_changes' is defined here`|Fails with<br>`TestActiveSupportExpectations#test_must_change [[...]/minitest-rails/test/rails/active_support/test_expectations.rb:8]:`<br>`Expected change from 0`<br><br>and gives the same warning

minitest-rails v6.1.0, Rails 6.1.7
| Ruby | Original test<br>(`counter = 0`) | Modified test<br>(`counter = 99`)
| - | - | -
| 3.1.2 | Passes with no warnings | Passes with no warnings
| 3.0.4 | Passes with no warnings | Passes with no warnings
| 2.7.6 | Passes with<br>`[...]/minitest-5.16.3/lib/minitest/spec.rb:26: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call`<br>`[...]/gems/activesupport-7.0.4/lib/active_support/testing/assertions.rb:175: warning: The called method 'assert_changes' is defined here`|Fails with<br>`TestActiveSupportExpectations#test_must_change [[...]/minitest-rails/test/rails/active_support/test_expectations.rb:8]:`<br>`Expected change from 0`<br><br>and gives the same warning

So it looks like it's (mostly?) related to Ruby 3.x parser changes?

**Comment by bobmazanec** [comment]
I'm relatively new to minitest{,-rails}, but am happy to keep learning and try helping.

That leads me to: What kind(s) of test/examples would you want to have caught this—especially since this uses a method in one gem (`infect_an_assertion`) to create a new method "based on" a method in another gem (`assert_changes`).

Said another way: which of those methods' 'implementation details' would you want to test?

I think that testing that `must_change` (et al?) actually raises `Minitest::Assertion` when appropriate—like [this minitest test/example](https://github.com/minitest/minitest/blob/b5565c0c7ab2ce5ecf306487cab0f2abe5377d55/test/minitest/test_minitest_spec.rb#L58-L64)—would be the most comprehensive. That is, testing `must_change`'s behavior, independent of how that is implemented/provided. I think this would also take
- Either duplicating things like minitest's [`assert_triggered`](https://github.com/minitest/minitest/blob/b5565c0c7ab2ce5ecf306487cab0f2abe5377d55/test/minitest/test_minitest_assertions.rb#L70) in this gem's tests
- Or separating them from [`TestMinitestAssertions`](https://github.com/minitest/minitest/blob/b5565c0c7ab2ce5ecf306487cab0f2abe5377d55/test/minitest/test_minitest_assertions.rb#L24) so they could be used here without all of that class's actual test cases.

I could also see an argument for mocking `assert_changes` to check that calling `must_change` passes expected arguments to it—in effect, testing `infect_an_assertion` and that mintiest-rails called it with the expected symbols, etc., but I'm not confident that would catch parser-change-caused-problems like this one.

Of course dev and/or CI changes to test against different Ruby versions is another story...

🤷‍♂️

**Comment by blowmage** [comment]
Hey Bob, thank you for reporting this issue. I haven't been ignoring it, but I was recently laid off and am kinda busy with dealing with that...

I did talk a bit with Ryan about this. I have successfully reproduced this, and am working on a fix. Will try to get something out soon.

**Comment by bobmazanec** [comment]
Thanks for the reply (and gem!); sorry to read of your layoff

No demand/expectation of a fix or schedule, of course—it's FOSS, right?

It's been kinda fun to dig in & see/learn some new stuff. Happy to help as able, but probably nowhere the "speed" of someone already familiar with all this when you/they get to focus on it.

**Comment by bobmazanec** [comment]
Fixed by minitest/minitest@87604fc
Released in [minitest v5.17.0](https://github.com/minitest/minitest/releases/tag/v5.17.0)

Thanks!

**Comment by blowmage** [comment]
ᕕ( ᐛ )ᕗ
